### ISSUE TITLE
`join` from outside a task takes a worker slot it never gave up

### ISSUE THREAD

**Issue by dritory** [rask-lang/rask#1346]
With a bounded worker count, joining from a caller that isn't a task (a test body, or `main`) takes one of the worker slots when the join returns, and keeps it. A spawned task can then starve.

```rask
import async.spawn

test "join then receive" {
    using Multitasking(workers: 1) {
        let (tx, rx) = Channel<i64>.buffered(1)
        let a = spawn(|| { })
        ensure a.detach()
        let b = spawn(|| {
            let _ = tx.send(7)
        })
        ensure b.detach()
        a.join()!
        let v = rx.receive()!
        assert v == 7
        b.join()!
    }
}
```

`rask test --sim --seeds 100` deadlocks by the second seed:
- main is waiting on the `receive`;
- task 2 is waiting on a worker slot, which main now holds.

In production this is a race that usually passes.

**Cause:** in `compiler/runtime/thread.c` (~210), `rask_task_slot_retake` calls `slot_take` whether or not the matching `slot_give` actually released a slot. A caller that isn't a task never had a slot, so it gives nothing back but still takes one.

**Fix:** `slot_give` reports whether it released a slot, and the join retakes only in that case.

Found by the sim-mode review (#625, PR #1337).
